mudslide/turbomole_model.py

```diff
@@ -0,0 +1,234 @@
+
+    def turbomole_init(self):
+        self.coord_path = Path(self.turbomole_dir)/"control"  
+        self.get_coords()
+
+    def get_coords(self):
+        coords = []
+        self.atom_order = []
+        self.mass = []
+
+        coord = subprocess.run(["sdg", "coord"], capture_output= True, text = True).stdout
+    
+        coord_list = coord.rstrip().split("\n")
+
+        for c in coord_list[1:]:
+            c_list = c.split()
+            self.atom_order.append(c_list[3])
+            coords.append([float(val) for val in c_list[:3]])
+            self.mass.append(3 * [masses[c_list[3]]])
+
+        self.ndim_ = 3 * len(coords)
+        self.X = np.array(coords, dtype=np.float64).reshape(self.ndim())
+        self.mass = np.array(self.mass, dtype=np.float64).reshape(self.ndim()) * amu_to_au
+
+
+    def update_coords(self, X):
+        self.coord_path = subprocess.run(["sdg", "-f", "coord"], capture_output=True, text=True).stdout.rstrip()
+        X = X.reshape((self.ndim() // 3, 3))
+        with open(self.coord_path, "r") as f:
+            lines = f.readlines()
+        regex = re.compile(r"\$coord\n|\$coord\s")
+        read_start = np.argwhere(np.array([bool(regex.match(line)) for line in lines]))[0][0]
+
+        current_line = read_start +1
+        for i, coord_list in enumerate(X):
+            lines[current_line] = "    {: .14f}      {: .14f}      {: .14f}      {}\n".format(
+                                  coord_list[0], coord_list[1], coord_list[2], self.atom_order[i]
+                    )
+            current_line += 1
+        
+        with open(self.coord_path, "w") as coord_file:
+            coord_file.write("".join(lines))            
+
+
+    def call_turbomole(self, outname="turbo.out"):
+        # Open file/run Turbomole
+        with open(outname, "w") as f:
+            for turbomole_module in self.turbomole_modules.values():
+                tur_output = subprocess.run(turbomole_module, stdout=f) 
+
+        # Parse results with Turboparse
+        with open(outname, "r") as f:
+            data_dict = turboparse.parse_turbo(f)
+
+        # Now add results to model
+    def call_turbomole(self, outname="turbo.out"):
+
+        with open(outname, "w") as f:
+            for turbomole_module in self.turbomole_modules.values():
+                tur_output = subprocess.run(turbomole_module, stdout=f) 
+
+        # Parse results with Turboparse
+        with open(outname, "r") as f:
+            data_dict = turboparse.parse_turbo(f)
+
+        # Now add results to model
+        energy = data_dict[self.turbomole_modules["gs_energy"]]["energy"]
+        self.energies = [energy]
+        parsed_gradients = data_dict[self.turbomole_modules["gs_grads"]]["gradient"]
+
+        # Check for presence of egrad turbomole module
+
+        if "egrad" in self.turbomole_modules.values():
+            # egrad updates to energy
+            parsed_energies = data_dict["egrad"]["excited_state"][0]["energy"]
```

Comment:
Does this line do anything?

Reply:
What if someone calls turbomole_model but doesn’t want to run egrad? The code wouldn’t currently run, but this allows for the possibility for someone else to do something else.  makes sense to you?

Reply:
I meant this one line in particular that defines the variable `parsed_energies`. The variable `parsed_energies` isn't used anywhere else in the file.

Reply:
you're right - removed line parsed_energies